addons/binding/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/handler/DeviceHandler.java

```diff
@@ -0,0 +1,149 @@
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.PointType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.icloud.internal.Address;
+import org.openhab.binding.icloud.internal.json.icloud.Content;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Patrik Gfeller
+ * @author Hans-Jörg Merk
+ */
+public class DeviceHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(DeviceHandler.class);
+    private int index;
+    private String deviceId;
+    private BridgeHandler bridge;
+    private LocationProvider locationProvider;
+
+    public DeviceHandler(@NonNull Thing thing, LocationProvider locationProvider) {
+        super(thing);
+
+        String uid = thing.getUID().toString();
+        index = Integer.parseInt(uid.substring(uid.lastIndexOf(":") + 1));
+
+        this.locationProvider = locationProvider;
+    }
+
+    public void update(ArrayList<Content> content) {
+        Content deviceData = getMyContent(content);
+        if (deviceData != null) {
+            deviceId = deviceData.getId();
+            updateStatus(ThingStatus.ONLINE);
+
+            updateState(BATTERYSTATUS, new StringType(deviceData.getBatteryStatus()));
+
+            Double batteryLevel = deviceData.getBatteryLevel();
+            if ((batteryLevel != null) && (batteryLevel != Double.NaN)) {
+                updateState(BATTERYLEVEL, new DecimalType(deviceData.getBatteryLevel() * 100));
+            }
+
+            if (deviceData.getLocation() != null) {
+                DecimalType latitude = new DecimalType(deviceData.getLocation().getLatitude());
+                DecimalType longitude = new DecimalType(deviceData.getLocation().getLongitude());
+                DecimalType accuracy = new DecimalType(deviceData.getLocation().getHorizontalAccuracy());
+
+                if (deviceData.getLocation().getTimeStamp() > 0) {
+                    Date javaDate = new Date(deviceData.getLocation().getTimeStamp());
+                    SimpleDateFormat javaDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssz");
+                    String lastUpdate = javaDateFormat.format(javaDate);
+
+                    updateState(LASTUPDATE, new DateTimeType(lastUpdate));
+                }
+
+                PointType location = new PointType(latitude, longitude);
+                Address address = bridge.getAddress(location);
```

Comment:
bridge.getAddress can return null (AddressLookupParser fails). Null is not checked so it seems to cause some problems. If address is null, address channels should be set to UNDEF.

2017-10-02 19:42:20.761 [ERROR] [binding.icloud.handler.BridgeHandler] - Index: 0, Size: 0
[Ljava.lang.StackTraceElement;@23e6335d
2017-10-02 19:42:20.768 [WARN ] [pse.smarthome.core.items.GenericItem] - failed notifying listener 'org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl@7a6d07b8' about state update of item Phone_Pauli_Address: null
java.lang.NullPointerException: null
	at org.mapdb.DataOutput2.writeUTF(DataOutput2.java:147) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.openhab.persistence.mapdb.internal.MapDBitemSerializer.serialize(MapDBitemSerializer.java:45) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.openhab.persistence.mapdb.internal.MapDBitemSerializer.serialize(MapDBitemSerializer.java:1) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.BTreeMap$NodeSerializer.serialize(BTreeMap.java:397) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.BTreeMap$NodeSerializer.serialize(BTreeMap.java:288) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.Store.serialize(Store.java:154) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.StoreWAL.update(StoreWAL.java:371) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.Caches$HashTable.update(Caches.java:270) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.EngineWrapper.update(EngineWrapper.java:63) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.BTreeMap.put2(BTreeMap.java:707) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.mapdb.BTreeMap.put(BTreeMap.java:643) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.openhab.persistence.mapdb.internal.MapDBPersistenceService.store(MapDBPersistenceService.java:161) [248:org.openhab.persistence.mapdb:1.11.0.201709280110]
	at org.openhab.core.persistence.internal.PersistenceServiceDelegate.store(PersistenceServiceDelegate.java:55) [188:org.openhab.core.compat1x:2.2.0.201709301515]
	at org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl.handleStateEvent(PersistenceManagerImpl.java:126) [113:org.eclipse.smarthome.core.persistence:0.9.0.201709260841]
	at org.eclipse.smarthome.core.persistence.internal.PersistenceManagerImpl.stateChanged(PersistenceManagerImpl.java:427) [113:org.eclipse.smarthome.core.persistence:0.9.0.201709260841]
	at org.eclipse.smarthome.core.items.GenericItem$1.run(GenericItem.java:229) [108:org.eclipse.smarthome.core:0.9.0.201709260841]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]

Reply:
Hi @paulianttila 
thank you for your feedback; if i look at the call stak it looks like the item value is loaded from persistence, not from the parser. Was this during start up of the binding?

I´ll check a null check to the code in question; but I suspect this is another issue,

Reply:
Thanks for the fix. Root cause was that iCloud binding set address channel to StringType(null) which cause the problem.  There is a separate issue about that mapdb openhab/openhab1-addons/issues/5285.

Reply:
I didn't test the fix (yet), but it might be that this fix doesn't completely fix the issue. As address itself is not null, but address class variables.

Reply:
Hi @paulianttila,

you´re right. I´ve added additional null checks for the properties as well.